### PR TITLE
Initialize resources on demand

### PR DESCRIPTION
**Pull request by davidbeckingsale**

### CODE REVIEW

File: src/umpire/resource/MemoryResourceTypes.hpp [hunk 1]
```diff
@@ -18,7 +23,46 @@ struct MemoryResourceTypeHash {
...
+  else if (resource == "UM") return MemoryResourceType::Unified;
+  else if (resource == "PINNED") return MemoryResourceType::Pinned;
+  else if (resource == "DEVICE_CONST") return MemoryResourceType::Constant;
+  else if (resource == "FILE") return MemoryResourceType::File;
+  else UMPIRE_ERROR("Unkown resource name: " << resource);
+}
+
+inline int resource_to_device_id(const std::string& resource) {
+  int device_id{0};
+  if (resource.find("::") != std::string::npos) {
```
Comment (davidbeckingsale) [inline src/umpire/resource/MemoryResourceTypes.hpp:61]: Could use # as the separator.

Comment (mcfadden8) [reply src/umpire/resource/MemoryResourceTypes.hpp:61]: @davidbeckingsale, what is left to do on this?  Is it just replay?  If so, I can work on getting replay to work.

**Comment by mcfadden8** [comment]
LGTM

**Comment by mcfadden8** [comment]
LGTM

**Review by mcfadden8 — APPROVED** [review]
Looks good!

**Comment by davidbeckingsale** [comment]
LGTM

**Review by kab163 — APPROVED** [review]

**Comment by davidbeckingsale** [comment]
LGTM

**Comment by davidbeckingsale** [comment]
LGTM

**Pull request merged**